src/pudl/transform/eia861.py

```diff
@@ -175,14 +434,176 @@
     ("WA", "Wahkiakurn", "Wahkiakum"),
 ], columns=["state", "eia_county", "fips_county"])
 
+BA_NAME_FIXES = pd.DataFrame([
+    ("Omaha Public Power District", 14127, "OPPD"),
+    ("Kansas City Power & Light Co", 10000, "KCPL"),
+    ("Toledo Edison Co", 18997, pd.NA),
+    ("Ohio Edison Co", 13998, pd.NA),
+    ("Cleveland Electric Illum Co", 3755, pd.NA),
+], columns=["balancing_authority_name_eia",
+            "balancing_authority_id_eia",
+            "balancing_authority_code_eia",
+            ]
+)
+
+CUSTOMER_CLASSES = [
+    "commercial",
+    "industrial",
+    "other",
+    "residential",
+    "total",
+    "transportation"
+]
+
+###############################################################################
+# EIA Form 861 Transform Helper functions
+###############################################################################
+
+
+def _filter_customer_cols(df, customer_classes):
+    regex = f"^({'_|'.join(customer_classes)}).*$"
+    return df.filter(regex=regex)
+
+
+def _filter_non_customer_cols(df, customer_classes):
+    regex = f"^(?!({'_|'.join(customer_classes)})).*$"
+    return df.filter(regex=regex)
+
+
+def _ba_code_backfill(df):
+    """
+    Backfill Balancing Authority Codes based on codes in later years.
+
+    Note:
+        The BA Code to ID mapping can change from year to year. If a Balancing Authority
+        is bought by another entity, the code may change, but the old EIA BA ID will be
+        retained.
+
+    Args:
+        ba_eia861 (pandas.DataFrame): The transformed EIA 861 Balancing
+            Authority dataframe (balancing_authority_eia861).
+
+    Returns:
+        pandas.DataFrame: The balancing_authority_eia861 dataframe, but with
+        many fewer NA values in the balancing_authority_code_eia column.
+
+    """
+    start_len = len(df)
+    start_nas = len(df.loc[df.balancing_authority_code_eia.isnull()])
+    logger.info(
+        f"Started with {start_nas} missing BA Codes out of {start_len} "
+        f"records ({start_nas/start_len:.2%})")
+    ba_ids = (
+        df[["balancing_authority_id_eia", "balancing_authority_code_eia", "report_date"]]
+        .drop_duplicates()
+        .sort_values(["balancing_authority_id_eia", "report_date"])
+    )
+    ba_ids["ba_code_filled"] = (
+        ba_ids.groupby("balancing_authority_id_eia")[
+            "balancing_authority_code_eia"]
+        .apply(lambda x: x.bfill())
+    )
```

Comment:
I'm not super familiar with `bfill` but why do you need to use a groupby and apply here? it seems like a simple `ba_ids["balancing_authority_code_eia"].fillna(method="bfill")` would suffice.

Reply:
What I'm doing here is trying to backfill values of `balancing_authority_code_eia` *within* a particular value of `balancing_authority_id_eia` -- we don't want to just backfill the code, because many different balancing authorities are mixed together here.  We want to take all records where, say, `balancing_authority_id_eia==15466` and ensure that the older ones, which have no code associated with them, are given whatever code the newer records associated with that ID have. (i.e. "PSCO").

Reply:
Sure. It still feels like something like this would work and would avoid the apply:

```
(ba_ids.groupby("balancing_authority_id_eia")["balancing_authority_code_eia"]
.fillna(method="bfill"))
```

Reply:
Huh, yeah it looks like you're correct. I haven't used it like that before.